# Invitae Cohort Analysis

By Jackie Petersen

---

In [209]:
# imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [210]:
# reading in the data
orders = pd.read_csv('orders.csv')
customers = pd.read_csv('customers.csv')

### Exploring the data

In [211]:
orders.head()

id  order_number  user_id              created
0  1709            36      344  2014-10-28 00:20:01
1  1406             7      608  2014-10-14 23:44:53
2  1716             6     2296  2014-10-28 17:47:07
3  1426             2     1225  2014-10-15 18:33:38
4  1415             6      797  2014-10-15 02:07:16

In [212]:
customers.head()

id              created
0  35410  2015-07-03 22:01:11
1  35417  2015-07-03 22:11:23
2  35412  2015-07-03 22:02:52
3  35413  2015-07-03 22:05:02
4  35424  2015-07-03 22:21:55

In [213]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27575 entries, 0 to 27574
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   id            27575 non-null  int64 
 1   order_number  27575 non-null  int64 
 2   user_id       27575 non-null  int64 
 3   created       27575 non-null  object
dtypes: int64(3), object(1)
memory usage: 861.8+ KB


In [214]:
customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25716 entries, 0 to 25715
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   id       25716 non-null  int64 
 1   created  25716 non-null  object
dtypes: int64(1), object(1)
memory usage: 401.9+ KB


> There are no NaN's in the data.

In [215]:
orders.sort_values(by='user_id')

id  order_number  user_id              created
26209      1             0        2  2014-05-16 00:13:50
25480      4             0        3  2014-05-21 20:40:05
26055      7             0        3  2014-05-21 23:00:14
26211      3             0        3  2014-05-21 20:39:23
26212      8             0        3  2014-05-22 20:51:22
...      ...           ...      ...                  ...
21551  28730             1    35964  2015-07-08 02:05:56
21507  28743             1    35976  2015-07-08 03:04:46
21588  28752             1    35977  2015-07-08 03:37:49
21638  28805             1    35990  2015-07-08 15:40:00
21601  28770             1    35994  2015-07-08 05:12:04

[27575 rows x 4 columns]

In [216]:
len(orders['user_id'].unique())

7067

In [217]:
len(orders['id'].unique())

27575

> In the orders dataset, the `'id'` column contains all unique values, where the `'user_id'` has duplicate values, perhaps for multiple orders from the same customers.

In [218]:
customers.sort_values(by='id')

id              created
1281   9483  2015-01-01 00:26:59
1072   9484  2015-01-01 00:51:40
1073   9485  2015-01-01 01:13:18
1074   9486  2015-01-01 01:57:02
1075   9487  2015-01-01 02:22:41
...     ...                  ...
491   35948  2015-07-07 23:20:54
531   35949  2015-07-07 23:23:04
434   35950  2015-07-07 23:25:42
444   35951  2015-07-07 23:29:58
516   35952  2015-07-07 23:34:39

[25716 rows x 2 columns]

In [219]:
len(customers['id'].unique())

25716

> The `'id'` column in the customers dataset has all unique values.

### Renaming the columns to differentiate before combining

In [220]:
orders.rename(columns={'id':'orders_id', 'created':'order_created'}, inplace=True)
orders.head()

orders_id  order_number  user_id        order_created
0       1709            36      344  2014-10-28 00:20:01
1       1406             7      608  2014-10-14 23:44:53
2       1716             6     2296  2014-10-28 17:47:07
3       1426             2     1225  2014-10-15 18:33:38
4       1415             6      797  2014-10-15 02:07:16

In [221]:
customers.rename(columns={'id':'user_id', 'created':'customer_created'}, inplace=True)
customers.head()

user_id     customer_created
0    35410  2015-07-03 22:01:11
1    35417  2015-07-03 22:11:23
2    35412  2015-07-03 22:02:52
3    35413  2015-07-03 22:05:02
4    35424  2015-07-03 22:21:55

### Combining the two datasets

> On the `'user_id'` column from the orders dataset and the `'id'` column on the customers dataset, now renamed `'user_id'`.

In [222]:
df_combined = orders.merge(customers, on='user_id')
df_combined

orders_id  order_number  user_id        order_created  \
0           8256             2    12654  2015-02-17 20:03:27   
1          12711             5    12654  2015-03-24 02:03:44   
2          11763             4    12654  2015-03-16 23:00:46   
3          21112             6    12654  2015-05-25 18:13:47   
4          22186             7    12654  2015-06-01 22:41:03   
...          ...           ...      ...                  ...   
17010       5214             1    11828  2015-01-19 01:17:24   
17011       5190             1    11209  2015-01-18 23:46:18   
17012       7689             1    13496  2015-02-11 03:44:58   
17013       7640             1    14980  2015-02-11 00:47:08   
17014       8281             1    12171  2015-02-17 22:32:19   

          customer_created  
0      2015-01-20 02:45:51  
1      2015-01-20 02:45:51  
2      2015-01-20 02:45:51  
3      2015-01-20 02:45:51  
4      2015-01-20 02:45:51  
...                    ...  
17010  2015-01-14 04:06:27  
17011  2015-01-11 18:14:36  
17012  2015-01-24 02:55:06  
17013  2015-02-08 20:18:50  
17014  2015-01-15 23:03:43  

[17015 rows x 5 columns]

In [223]:
len(df_combined['user_id'].unique())

5356

> There are 5356 unique customers now in the combined dataset.

In [224]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17015 entries, 0 to 17014
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   orders_id         17015 non-null  int64 
 1   order_number      17015 non-null  int64 
 2   user_id           17015 non-null  int64 
 3   order_created     17015 non-null  object
 4   customer_created  17015 non-null  object
dtypes: int64(3), object(2)
memory usage: 797.6+ KB


### Converting the `'order_created'` and `'customer_created'` columns to datetime format

In [225]:
df_combined['order_created'] = pd.to_datetime(df_combined['order_created'])
df_combined['customer_created'] = pd.to_datetime(df_combined['customer_created'])

In [226]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17015 entries, 0 to 17014
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   orders_id         17015 non-null  int64         
 1   order_number      17015 non-null  int64         
 2   user_id           17015 non-null  int64         
 3   order_created     17015 non-null  datetime64[ns]
 4   customer_created  17015 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 797.6 KB


### Converting the UTC timezone to US/Eastern

In [227]:
df_combined['order_created'] = df_combined['order_created'].dt.tz_localize(tz='UTC').dt.tz_convert(tz='US/Eastern')
df_combined['customer_created'] = df_combined['customer_created'].dt.tz_localize(tz='UTC').dt.tz_convert(tz='US/Eastern')

In [228]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17015 entries, 0 to 17014
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype                     
---  ------            --------------  -----                     
 0   orders_id         17015 non-null  int64                     
 1   order_number      17015 non-null  int64                     
 2   user_id           17015 non-null  int64                     
 3   order_created     17015 non-null  datetime64[ns, US/Eastern]
 4   customer_created  17015 non-null  datetime64[ns, US/Eastern]
dtypes: datetime64[ns, US/Eastern](2), int64(3)
memory usage: 797.6 KB


In [229]:
df_combined.head()

orders_id  order_number  user_id             order_created  \
0       8256             2    12654 2015-02-17 15:03:27-05:00   
1      12711             5    12654 2015-03-23 22:03:44-04:00   
2      11763             4    12654 2015-03-16 19:00:46-04:00   
3      21112             6    12654 2015-05-25 14:13:47-04:00   
4      22186             7    12654 2015-06-01 18:41:03-04:00   

           customer_created  
0 2015-01-19 21:45:51-05:00  
1 2015-01-19 21:45:51-05:00  
2 2015-01-19 21:45:51-05:00  
3 2015-01-19 21:45:51-05:00  
4 2015-01-19 21:45:51-05:00

### Stripping the extra timezone info off the end of the datetime columns, then converting back to datetime format

In [230]:
df_combined['order_created'] = df_combined['order_created'].dt.strftime('%Y-%m-%d %H:%M:%S')
df_combined['customer_created'] = df_combined['customer_created'].dt.strftime('%Y-%m-%d %H:%M:%S')
df_combined['order_created'] = pd.to_datetime(df_combined['order_created'])
df_combined['customer_created'] = pd.to_datetime(df_combined['customer_created'])

In [231]:
df_combined.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17015 entries, 0 to 17014
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   orders_id         17015 non-null  int64         
 1   order_number      17015 non-null  int64         
 2   user_id           17015 non-null  int64         
 3   order_created     17015 non-null  datetime64[ns]
 4   customer_created  17015 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(3)
memory usage: 797.6 KB


In [232]:
df_combined.head()

orders_id  order_number  user_id       order_created    customer_created
0       8256             2    12654 2015-02-17 15:03:27 2015-01-19 21:45:51
1      12711             5    12654 2015-03-23 22:03:44 2015-01-19 21:45:51
2      11763             4    12654 2015-03-16 19:00:46 2015-01-19 21:45:51
3      21112             6    12654 2015-05-25 14:13:47 2015-01-19 21:45:51
4      22186             7    12654 2015-06-01 18:41:03 2015-01-19 21:45:51

### Creating a column with the difference in days of the customer creation and the order creation

In [233]:
df_combined['time_difference'] = df_combined['order_created'] - df_combined['customer_created']
df_combined.head()

orders_id  order_number  user_id       order_created    customer_created  \
0       8256             2    12654 2015-02-17 15:03:27 2015-01-19 21:45:51   
1      12711             5    12654 2015-03-23 22:03:44 2015-01-19 21:45:51   
2      11763             4    12654 2015-03-16 19:00:46 2015-01-19 21:45:51   
3      21112             6    12654 2015-05-25 14:13:47 2015-01-19 21:45:51   
4      22186             7    12654 2015-06-01 18:41:03 2015-01-19 21:45:51   

    time_difference  
0  28 days 17:17:36  
1  63 days 00:17:53  
2  55 days 21:14:55  
3 125 days 16:27:56  
4 132 days 20:55:12

In [234]:
df_combined['time_difference'] = df_combined['time_difference'].dt.days
df_combined.head()

orders_id  order_number  user_id       order_created    customer_created  \
0       8256             2    12654 2015-02-17 15:03:27 2015-01-19 21:45:51   
1      12711             5    12654 2015-03-23 22:03:44 2015-01-19 21:45:51   
2      11763             4    12654 2015-03-16 19:00:46 2015-01-19 21:45:51   
3      21112             6    12654 2015-05-25 14:13:47 2015-01-19 21:45:51   
4      22186             7    12654 2015-06-01 18:41:03 2015-01-19 21:45:51   

   time_difference  
0               28  
1               63  
2               55  
3              125  
4              132

### Stripping the timestamp from the `'order_created'` and `'customer_created'` columns and adding columns with the week number for both

In [235]:
df_combined['order_week'] = df_combined['order_created'].dt.isocalendar()['week']
df_combined['customer_week'] = df_combined['customer_created'].dt.isocalendar()['week']
df_combined['order_created'] = df_combined['order_created'].dt.date
df_combined['customer_created'] = df_combined['customer_created'].dt.date
df_combined.head()

orders_id  order_number  user_id order_created customer_created  \
0       8256             2    12654    2015-02-17       2015-01-19   
1      12711             5    12654    2015-03-23       2015-01-19   
2      11763             4    12654    2015-03-16       2015-01-19   
3      21112             6    12654    2015-05-25       2015-01-19   
4      22186             7    12654    2015-06-01       2015-01-19   

   time_difference  order_week  customer_week  
0               28           8              4  
1               63          13              4  
2               55          12              4  
3              125          22              4  
4              132          23              4

In [247]:
df_combined['weekly_difference'] = df_combined['order_week'] - df_combined['customer_week']
df_combined.head()

orders_id  order_number  user_id order_created customer_created  \
0       8256             2    12654    2015-02-17       2015-01-19   
1      12711             5    12654    2015-03-23       2015-01-19   
2      11763             4    12654    2015-03-16       2015-01-19   
3      21112             6    12654    2015-05-25       2015-01-19   
4      22186             7    12654    2015-06-01       2015-01-19   

   time_difference  order_week  customer_week  weekly_difference  
0               28           8              4                  4  
1               63          13              4                  9  
2               55          12              4                  8  
3              125          22              4                 18  
4              132          23              4                 19

In [250]:
df_combined.sort_values(by='customer_week', ascending=False)

orders_id  order_number  user_id order_created customer_created  \
16251      28322             1    35755    2015-07-06       2015-07-06   
16305      28680             2    35904    2015-07-07       2015-07-07   
16307      28560             1    35895    2015-07-07       2015-07-07   
16255      28404             1    35736    2015-07-06       2015-07-06   
16313      28661             1    35910    2015-07-07       2015-07-07   
...          ...           ...      ...           ...              ...   
7968        7113             2     9508    2015-02-04       2015-01-01   
111        18924            27     9560    2015-05-09       2015-01-02   
110        22508            35     9560    2015-06-03       2015-01-02   
109        24913            41     9560    2015-06-17       2015-01-02   
1332       15130            15     9549    2015-04-13       2015-01-02   

       time_difference  order_week  customer_week  weekly_difference  
16251                0          28             28                  0  
16305                0          28             28                  0  
16307                0          28             28                  0  
16255                0          28             28                  0  
16313                0          28             28                  0  
...                ...         ...            ...                ...  
7968                34           6              1                  5  
111                126          19              1                 18  
110                151          23              1                 22  
109                165          25              1                 24  
1332               100          16              1                 15  

[17015 rows x 9 columns]